ouroboros-consensus-test/bench-mempool/Main.hs

```diff
@@ -18,37 +33,116 @@ import           Bench.Consensus.MempoolWithMockedLedgerItf
 import qualified Ouroboros.Consensus.Mempool.Capacity as Mempool
 
 main :: IO ()
-main = defaultMain [ benchmarkJustAddingTransactions ]
+main = do
+    let csvFilePath = "mempool-benchmarks.csv"
+    runBenchmarks csvFilePath
+    rawValues <- parseBenchmarkResults csvFilePath
+    convertCsvRowsToJsonObjects rawValues "mempool-benchmarks.json"
   where
-    benchmarkJustAddingTransactions =
-        bgroup "Just adding transactions" $
-            fmap benchAddNTxs [1_000, 10_000, 100_000, 1_000_000]
+    runBenchmarks csvFilePath = do
+        opts <- parseOptions benchIngredients benchmarkJustAddingTransactions
+        let opts' = changeOption (Just . fromMaybe (CsvPath csvFilePath)) opts
+        case tryIngredients benchIngredients opts' benchmarkJustAddingTransactions of
+          Nothing               -> exitFailure
+          Just    runIngredient -> do
+            success <- runIngredient
+            unless success exitFailure
+      where
+        benchmarkJustAddingTransactions =
+            bgroup "Just adding" $
+                fmap benchAddNTxs [10_000, 1_000_000]
+          where
+            benchAddNTxs n =
+                env (let txs = mkNTryAddTxs n in fmap (, txs) (openMempoolWithCapacityFor txs))
+                -- The irrefutable pattern was added to avoid the
+                --
+                -- > Unhandled resource. Probably a bug in the runner you're using.
+                --
+                -- error reported here https://hackage.haskell.org/package/tasty-bench-0.3.3/docs/Test-Tasty-Bench.html#v:env
+                (\ ~(mempool, txs) -> bgroup (show n <> " transactions") [
+                    bench    "benchmark"     $ nfIO $ run        mempool txs
+                  , testCase "test"          $        testAddTxs mempool txs
+                  , testCase "txs length"    $ length txs @?= n
+                  ]
+                )
+              where
+                testAddTxs mempool txs = do
+                    run mempool txs
+                    mempoolTxs <- getTxs mempool
+                    mempoolTxs @?= txsAddedInCmds txs
+
+    parseBenchmarkResults csvFilePath = do
```

Comment:
cassava is already in the dependencies of ouroboros-network so it won't add much and will be more robust than hand-made parsing.

Reply:
I think if you do 
 
```
data BenchPoint = ...
    deriving (Generic, ToJSON, FromRow)
```

converting from CSV to JSON can boil down to a couple lines

Reply:
We need to do some picoseconds conversions and processing of the benchmark names, so I'd rather avoid this generic instance.

Reply:
> cassava is already in the dependencies of ouroboros-network so it won't add much and will be more robust than hand-made parsing.

Oh, good point. Let me try that.

Reply:
Done https://github.com/input-output-hk/ouroboros-network/pull/4422/commits/12962af5d3e7fab87e1565969aa061182d68526e